File: Geometry/PointInPolygon.cs
using NetTopologySuite.Geometries;

namespace PointInPolygonTest.Geometry
{
    public static class PointInPolygon
{
    public static bool Contains(Polygon polygon, Point p)
    {
        var coord = new Coordinate(p.Coordinate.X,  p.Coordinate.Y);
        
        if (!polygon.EnvelopeInternal.Contains(coord))
            return false;
        
        var pointInsideHole = polygon.Holes.Any(hole => Winding(hole.Coordinates, coord, true));
        if (pointInsideHole) return false;
        
        var windingExterior = Winding(polygon.ExteriorRing.Coordinates, coord, false);
        
        
        return windingExterior;
    }

    // -------------------------
    // 1. WINDING (non-zero rule)
    // -------------------------
    private static bool Winding(Coordinate[] ring, Coordinate p, bool isHole)
    {
        var wn = 0;
        var n = ring.Length;

        for (var i = 0; i < n; i++)
        {
            var a = ring[i];
            var b = ring[(i + 1) % n];

            if (IsSamePoint(a, p) || IsSamePoint(b, p) || OnSegment(a, b, p)) return !isHole;

            //пропускаем горизонтальные ребра
            if (Math.Abs(a.Y - b.Y) < MathConstants.E) continue;
            
            if (a.Y <= p.Y)
            {
                if (b.Y > p.Y && IsLeft(a, b, p))
                    wn++;
            }
            else
            {
                if (b.Y <= p.Y && !IsLeft(a, b, p))
                    wn--;
            }
            Console.WriteLine(wn);
        }

        return wn != 0;
    }
    //если площадь параллелограма на двух векторах меньше нуля значит угол между ними больше 180
    private static bool IsLeft(Coordinate a, Coordinate b, Coordinate p)
    {
        var ab = new Coordinate(b.X - a.X, b.Y - a.Y);
        var ap = new Coordinate(p.X - a.X, p.Y - a.Y);
        
        return Cross(ab, ap) > 0;
    }

    private static bool IsSamePoint(Coordinate a, Coordinate b)
    {
        return Math.Abs(a.X - b.X) < MathConstants.E && Math.Abs(a.Y - b.Y) <  MathConstants.E;
    }

    private static bool OnSegment(Coordinate a, Coordinate b, Coordinate p)
    {
        var ab = new Coordinate(b.X - a.X, b.Y - a.Y);
        var ap = new Coordinate(p.X - a.X, p.Y - a.Y);
        //если ab и ap не параллельные вектора
        if (Math.Abs(Cross(ab, ap)) > MathConstants.E) return false;
        //ab || ap проверяем лежит ли p в границах отрезка ab
        return p.X >= Math.Min(a.X, b.X) - MathConstants.E && p.X <= Math.Max(a.X, b.X) + MathConstants.E &&
               p.Y >= Math.Min(a.Y, b.Y) - MathConstants.E && p.Y <= Math.Max(a.Y, b.Y) + MathConstants.E;
    }
    
    //векторное произведение
    private static double Cross(Coordinate a, Coordinate b)
    {
        return a.X * b.Y - a.Y * b.X;
    }
}
}

File: src/main/java/com/example/demo/geometry/PointInPolygon.java
package com.example.demo.geometry;

import org.locationtech.jts.geom.*;

import java.util.List;

public class PointInPolygon {
    private static final double EPSILON = 1e-9;

    public static boolean contains(Polygon polygon, Point point) {
        Coordinate coord = point.getCoordinate();

        if (!polygon.getEnvelopeInternal().contains(coord)) {
            return false;
        }

        for (int i = 0; i < polygon.getNumInteriorRing(); i++) {
            if (winding(polygon.getInteriorRingN(i).getCoordinates(), coord, true)) {
                return false;
            }
        }

        return winding(polygon.getExteriorRing().getCoordinates(), coord, false);
    }

    private static boolean winding(Coordinate[] ring, Coordinate p, boolean isHole) {
        int wn = 0;
        int n = ring.length;

        for (int i = 0; i < n; i++) {
            Coordinate a = ring[i];
            Coordinate b = ring[(i + 1) % n];

            if (isSamePoint(a, p) || isSamePoint(b, p) || onSegment(a, b, p)) {
                return !isHole;
            }

            if (Math.abs(a.getY() - b.getY()) < EPSILON) {
                continue;
            }

            if (a.getY() <= p.getY()) {
                if (b.getY() > p.getY() && isLeft(a, b, p)) {
                    wn++;
                }
            } else {
                if (b.getY() <= p.getY() && !isLeft(a, b, p)) {
                    wn--;
                }
            }
        }

        return wn != 0;
    }

    private static boolean isLeft(Coordinate a, Coordinate b, Coordinate p) {
        double abX = b.getX() - a.getX();
        double abY = b.getY() - a.getY();
        double apX = p.getX() - a.getX();
        double apY = p.getY() - a.getY();

        return cross(abX, abY, apX, apY) > 0;
    }

    private static boolean isSamePoint(Coordinate a, Coordinate b) {
        return Math.abs(a.getX() - b.getX()) < EPSILON &&
                Math.abs(a.getY() - b.getY()) < EPSILON;
    }

    private static boolean onSegment(Coordinate a, Coordinate b, Coordinate p) {
        double abX = b.getX() - a.getX();
        double abY = b.getY() - a.getY();
        double apX = p.getX() - a.getX();
        double apY = p.getY() - a.getY();

        if (Math.abs(cross(abX, abY, apX, apY)) > EPSILON) {
            return false;
        }

        return p.getX() >= Math.min(a.getX(), b.getX()) - EPSILON &&
                p.getX() <= Math.max(a.getX(), b.getX()) + EPSILON &&
                p.getY() >= Math.min(a.getY(), b.getY()) - EPSILON &&
                p.getY() <= Math.max(a.getY(), b.getY()) + EPSILON;
    }

    private static double cross(double ax, double ay, double bx, double by) {
        return ax * by - ay * bx;
    }
}